### PR TITLE
[GLUTEN-5620][CORE] Simplify Decimal process logic

### PR DESCRIPTION
**Pull request by baibaichen**
## What changes were proposed in this pull request?

This PR trys to unify decimal process logic for https://github.com/apache/spark/pull/36698.

|                                    | Before 36698                                                 | After 36698                                                  |
| ---------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| `PromotePrecision`                 | Actually, it is a placeholder, it dose nothing at runtime.   | Removed                                                      |
| `CheckOverflow`                    | It calls `Decimal::toPrecision` to change decimal's type to result type. | Merge `Decimal::toPrecision` into arithmetic `eval` and code-gen code path, and still use it in `RowEncoder` |
| `Add`, `Divide`,  `Multiply`, etc. | `dataType` is  not actual result type.                       | `dataType` is  actual result type.                           |

### How

1. Don't transform `CheckOverflow` when its child is decimal `BinaryArithmetic` in `ExpressionConverter`:

```scala
      case CheckOverflow(b: BinaryArithmetic, decimalType, _)
          if !BackendsApiManager.getSettings.transformCheckOverflow &&
             DecimalArithmeticUtil.isDecimalArithmetic(b) =>
        DecimalArithmeticUtil.checkAllowDecimalArithmetic()
        val leftChild =
          replaceWithExpressionTransformerInternal(b.left, attributeSeq, expressionsMap)
        val rightChild =
          replaceWithExpressionTransformerInternal(b.right, attributeSeq, expressionsMap)
        DecimalArithmeticExpressionTransformer(
          getAdnCheckSubstraitName(b, expressionsMap),
          leftChild,
          rightChild,
          decimalType,
          b)
```

2. Automatically add native  `CheckOverflow`  when doing decimal `BinaryArithmetic` at backend, see `FunctionParserBinaryArithmetic::parese`

> [!NOTE]
>
> For the compatibility of the velox backed, `BackendSettingsApi::transformCheckOverflow` is introduced, which makes  `ExpressionConverter` a bit complicated.
>
> Hope we can remove this option and make code more clean, once velox has same logic. For example, we can remove `DecimalArithmeticUtil::getResultTypeForOperation`, since we can get it  either from `CheckOverflow` or decimal arithmetic expression itself.
>
> ```scala
>    val resultType = DecimalArithmeticUtil.getResultTypeForOperation(
>       DecimalArithmeticUtil.getOperationType(b),
>          DecimalArithmeticUtil
>            .getResultType(leftChild)
>            .getOrElse(left.dataType.asInstanceOf[DecimalType]),
>          DecimalArithmeticUtil
>            .getResultType(rightChild)
>            .getOrElse(right.dataType.asInstanceOf[DecimalType])
>        )
>    ```

(Fixes: \#5620)

## How was this patch tested?

Existed UT

### CODE REVIEW

**Comment by loneylee** [comment]
LGTM

**Comment by zhouyuan** [comment]
@jinchengchenghh can you please help to take a look on this patch?

**Comment by jinchengchenghh** [comment]
For velox backend, it check overflow in native binary arithmetic too, can you try to change the code to not transform checkoverflow?

**Comment by jinchengchenghh** [comment]
I think we can also remove `getResultTypeForOperation`

File: gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala [hunk 1]
```diff
@@ -510,42 +493,71 @@ object ExpressionConverter extends SQLConfHelper with Logging {
...
-          DecimalArithmeticUtil.getOperationType(b),
-          DecimalArithmeticUtil
-            .getResultType(leftChild)
-            .getOrElse(left.dataType.asInstanceOf[DecimalType]),
-          DecimalArithmeticUtil
-            .getResultType(rightChild)
-            .getOrElse(right.dataType.asInstanceOf[DecimalType])
-        )
+      case CheckOverflow(b: BinaryArithmetic, decimalType, _)
+          if DecimalArithmeticUtil.isDecimalArithmetic(b) &&
```
Comment (jinchengchenghh) [inline gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:498]: Can you change the condition to `BackendsApiManager.getSettings.dontTransformCheckOverflow && DecimalArithmeticUtil.isDecimalArithmetic(b)`, the first condition is easier to check

File: gluten-core/src/main/scala/org/apache/gluten/utils/DecimalArithmeticUtil.scala [hunk 2]
```diff
@@ -159,56 +163,30 @@ object DecimalArithmeticUtil {
...
-        } else {
-          (left, right)
-        }
-      } else {
-        (left, right)
-      }
+
+    if (!BackendsApiManager.getSettings.rescaleDecimalIntegralExpression()) {
+      (left, right)
+    } else if (!isPromoteCast(left) && isPromoteCastIntegral(right)) {
```
[the remark this answers is not in the document]
Comment (baibaichen) [reply gluten-core/src/main/scala/org/apache/gluten/utils/DecimalArithmeticUtil.scala:184]: done.

**Comment by baibaichen** [comment]
> For velox backend, it check overflow in native binary arithmetic too, can you try to change the code to not transform checkoverflow?

@jinchengchenghh would you please create followup pr to remove  `transformCheckOverflow` option?  I add this option for the compatibility, I am not familiar with velox backend.

**Comment by jinchengchenghh** [comment]
ok @baibaichen

**Review by jinchengchenghh — APPROVED** [review]

**Pull request merged**